# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [255]:
# !pip install -U scipy

In [256]:
# import numpy and pandas
import numpy as np 
import pandas as pd 

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [257]:
# Your code here:
salaries = pd.read_csv('Salaries.csv') 
salaries

C:\Users\chech\AppData\Local\Temp\ipykernel_24880\1511378093.py:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  salaries = pd.read_csv('Salaries.csv')


Id       EmployeeName  \
0            1     NATHANIEL FORD   
1            2       GARY JIMENEZ   
2            3     ALBERT PARDINI   
3            4  CHRISTOPHER CHONG   
4            5    PATRICK GARDNER   
...        ...                ...   
148649  148650      Roy I Tillery   
148650  148651       Not provided   
148651  148652       Not provided   
148652  148653       Not provided   
148653  148654          Joe Lopez   

                                              JobTitle       BasePay  \
0       GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY     167411.18   
1                      CAPTAIN III (POLICE DEPARTMENT)     155966.02   
2                      CAPTAIN III (POLICE DEPARTMENT)     212739.13   
3                 WIRE ROPE CABLE MAINTENANCE MECHANIC       77916.0   
4         DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)      134401.6   
...                                                ...           ...   
148649                                       Custodian          0.00   
148650                                    Not provided  Not Provided   
148651                                    Not provided  Not Provided   
148652                                    Not provided  Not Provided   
148653                      Counselor, Log Cabin Ranch          0.00   

         OvertimePay      OtherPay      Benefits   TotalPay  TotalPayBenefits  \
0                0.0     400184.25           NaN  567595.43         567595.43   
1          245131.88     137811.38           NaN  538909.28         538909.28   
2          106088.18       16452.6           NaN  335279.91         335279.91   
3           56120.71      198306.9           NaN  332343.61         332343.61   
4             9737.0     182234.59           NaN  326373.19         326373.19   
...              ...           ...           ...        ...               ...   
148649          0.00          0.00          0.00       0.00              0.00   
148650  Not Provided  Not Provided  Not Provided       0.00              0.00   
148651  Not Provided  Not Provided  Not Provided       0.00              0.00   
148652  Not Provided  Not Provided  Not Provided       0.00              0.00   
148653          0.00       -618.13          0.00    -618.13           -618.13   

        Year  Notes         Agency Status  
0       2011    NaN  San Francisco    NaN  
1       2011    NaN  San Francisco    NaN  
2       2011    NaN  San Francisco    NaN  
3       2011    NaN  San Francisco    NaN  
4       2011    NaN  San Francisco    NaN  
...      ...    ...            ...    ...  
148649  2014    NaN  San Francisco     PT  
148650  2014    NaN  San Francisco    NaN  
148651  2014    NaN  San Francisco    NaN  
148652  2014    NaN  San Francisco    NaN  
148653  2014    NaN  San Francisco     PT  

[148654 rows x 13 columns]

Examine the `salaries` dataset using the `head` function below.

In [258]:
# Your code here:
salaries.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [259]:
# Your code here:

In [260]:
salaries.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [261]:
# Your code here
salaries.shape

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [262]:
# Your code here
salaries.drop(columns=['Notes'], axis=1, inplace=True)


Check out what are the possible values of the column "Status".

In [263]:
# Your code here
salaries['Status'].value_counts(dropna=False)

NaN    110535
FT      22334
PT      15785
Name: Status, dtype: int64

In [264]:
salaries['Status'].unique()

array([nan, 'PT', 'FT'], dtype=object)

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [265]:
salaries1 = salaries.copy()

In [266]:
# Your code 
salaries1 = salaries[salaries['Status'].notna()]

In [267]:
# Your code here
salaries1.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [268]:
# Your code here
salaries = salaries1

In [269]:
salaries.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


Overtimepay, otherpay and Benefits must be int64. not objects.

In [270]:
# Your code here
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer   257340.0   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department   302068.0   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.0  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.0    82313.7  96570.66  339653.70         436224.36  2014   
110534         0.0   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.0   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

Do any type conversions and reset the index.

In [271]:
# Your code here
salaries.loc[:,['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']] = salaries[['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']].astype(float)


C:\Users\chech\AppData\Local\Temp\ipykernel_24880\1260066844.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  salaries.loc[:,['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']] = salaries[['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']].astype(float)
C:\Users\chech\AppData\Local\Temp\ipykernel_24880\1260066844.py:2: DeprecationWarning: In a future version, `df.iloc[:, i] = newvals` will attempt to set the values inplace instead of always setting a new array. To retain the old behavior, use either `df[df.columns[i]] = newvals` or, if columns are non-unique, `df.isetitem(i, newvals)`
  salaries.loc[:,['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']] = salaries[['BasePay', 'OvertimePay', 'OtherPay', 'Benefits']].astype(float)


In [272]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

        OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531         0.00  342802.63  38780.04  471952.64         510732.68  2014   
110532     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533         0.00   82313.70  96570.66  339653.70         436224.36  2014   
110534         0.00   19266.72  91302.46  326716.76         418019.22  2014   
110535         0.00   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

In [273]:
# Your code here
salaries.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  float64
 4   OvertimePay       38119 non-null  float64
 5   OtherPay          38119 non-null  float64
 6   Benefits          38119 non-null  float64
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(6), int64(2), object(4)
memory usage: 3.8+ MB


In [274]:
salaries = salaries.reset_index()

In [275]:
salaries = salaries.drop(columns=['index'])
salaries

Id          EmployeeName                     JobTitle    BasePay  \
0      110532           David Shinn               Deputy Chief 3  129150.01   
1      110533            Amy P Hart            Asst Med Examiner  318835.49   
2      110534  William J Coaker Jr.     Chief Investment Officer  257340.00   
3      110535        Gregory P Suhr              Chief of Police  307450.04   
4      110536  Joanne M Hayes-White       Chief, Fire Department  302068.00   
...       ...                   ...                          ...        ...   
38114  148646      Carolyn A Wilson    Human Services Technician       0.00   
38115  148648        Joann Anderson  Communications Dispatcher 2       0.00   
38116  148649           Leon Walker                    Custodian       0.00   
38117  148650         Roy I Tillery                    Custodian       0.00   
38118  148654             Joe Lopez   Counselor, Log Cabin Ranch       0.00   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0             0.00  342802.63  38780.04  471952.64         510732.68  2014   
1         10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2             0.00   82313.70  96570.66  339653.70         436224.36  2014   
3             0.00   19266.72  91302.46  326716.76         418019.22  2014   
4             0.00   24165.44  91201.66  326233.44         417435.10  2014   
...            ...        ...       ...        ...               ...   ...   
38114         0.00       0.00      0.00       0.00              0.00  2014   
38115         0.00       0.00      0.00       0.00              0.00  2014   
38116         0.00       0.00      0.00       0.00              0.00  2014   
38117         0.00       0.00      0.00       0.00              0.00  2014   
38118         0.00    -618.13      0.00    -618.13           -618.13  2014   

              Agency Status  
0      San Francisco     PT  
1      San Francisco     FT  
2      San Francisco     PT  
3      San Francisco     FT  
4      San Francisco     FT  
...              ...    ...  
38114  San Francisco     PT  
38115  San Francisco     PT  
38116  San Francisco     PT  
38117  San Francisco     PT  
38118  San Francisco     PT  

[38119 rows x 12 columns]

In [276]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [277]:
# Your code here
((salaries["BasePay"]  + salaries["OvertimePay"] + salaries["OtherPay"] + salaries["Benefits"]) == salaries["TotalPayBenefits"])


0         True
1        False
2         True
3        False
4         True
         ...  
38114     True
38115     True
38116     True
38117     True
38118     True
Length: 38119, dtype: bool

What is the percetage of employees for which the previous assumption is not True?

In [278]:
# Your code here:
((salaries["BasePay"]  + salaries["OvertimePay"] + salaries["OtherPay"] + salaries["Benefits"]) == salaries["TotalPayBenefits"]).value_counts()/len(salaries)*100

True     73.708649
False    26.291351
dtype: float64

In [279]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  
0  San Francisco     PT  
1  San Francisco     FT  
2  San Francisco     PT  
3  San Francisco     FT  
4  San Francisco     FT

There are different departments in the city. List all departments and the count of employees in each department.

In [280]:
# Your code here:

job_titles = salaries.groupby(['JobTitle']).agg({'EmployeeName':'count'})
job_titles = job_titles.rename(columns={'EmployeeName': 'Employees'})
job_titles

Employees
JobTitle                                 
ACPO,JuvP, Juv Prob (SFERS)             1
ASR Senior Office Specialist           22
ASR-Office Assistant                   15
Account Clerk                          93
Accountant I                            2
...                                   ...
Wire Rope Cable Maint Sprv              1
Worker's Comp Supervisor 1              6
Worker's Compensation Adjuster         26
X-Ray Laboratory Aide                  35
Youth Comm Advisor                      1

[996 rows x 1 columns]

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [281]:
# Your code here: (compute the "Hourly_Wage")
salaries['Hourly_Wage'] = salaries['TotalPayBenefits']/50/40
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [282]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)
FT_mean = salaries[salaries['Status']== 'FT']['Hourly_Wage'].mean()
FT_mean

69.26420640727143

In [283]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees


In [284]:
sample = salaries[salaries['Status']== 'FT']['Hourly_Wage'].to_numpy()
sample

array([239.826105, 209.00961 , 208.71755 , ...,  20.595055,  18.337755,
        15.986405])

In [285]:
t = (np.mean(sample) - 75)/(np.std(sample, ddof=1)/np.sqrt(len(sample)))
print("The value of the statistic is: ",t)

The value of the statistic is:  -35.80631941460526


In [286]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
lower_critical_value, upper_critical_value = scipy.stats.t.ppf((0.05/2), df=len(sample)-1), scipy.stats.t.ppf(1-(0.05/2), df=len(sample)-1)
print("The lower critical value is {:.3f}".format(lower_critical_value))
print("The upper critical value is {:.3f}".format(upper_critical_value))

The lower critical value is -1.960
The upper critical value is 1.960


Our statistic is not in between the critical values, therefore we reject the null hypothesis.

In [287]:
# Method 2: Use the p-value method.
# Your code here:
scipy.stats.t.cdf(t, df=len(sample)-1)

2.2589156169197237e-273

The area is not bigger than $\alpha$ and smaller than $1-\alpha$, therefore we reject the null hypothesis.

In [288]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
from scipy.stats import ttest_1samp
stat, pval = ttest_1samp(sample, popmean = 75, alternative = "two-sided")


In [289]:
stat, pval

(-35.80631941460526, 4.517831233839447e-273)

All the methods are in agreement.

Are all the methods in agreement?

We are also curious about salaries in the police force. The chief of police in San Francisco claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [290]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [291]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees
sample1 = salaries[salaries['JobTitle'].str.contains('Police')]['TotalPayBenefits'].to_numpy()
sample1

array([4.1801922e+05, 3.0174981e+05, 2.8911495e+05, ..., 2.1239000e+02,
       1.3478000e+02, 7.0370000e+01])

In [292]:
# sample1 = salaries[salaries['JobTitle'].apply(lambda x:'Police' if 'Police' in x else x)=='Police']['Hourly_Wage'].to_numpy()

In [293]:
t = (np.mean(sample1) - 86000)/(np.std(sample1, ddof=1)/np.sqrt(len(sample1)))
print("The value of the statistic is: ",t)

The value of the statistic is:  50.252984742109945


In [294]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 
critical_value = scipy.stats.t.ppf(0.05, df=len(sample1)-1)
print("The critical value is {:.3f}".format(critical_value))

The critical value is -1.646


In [295]:
# Method 2: Use the p-value method.
# Your code here:
scipy.stats.t.cdf(t, df=len(sample1)-1)

1.0

In our case:

$$ 1 > \alpha=0.05$$

Therefore, we **accept the null hypothesis** which is the same conclussion that we got with the other two methods. Great!!

In [296]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
stat, pval = ttest_1samp(sample1, popmean = 86000, alternative = "less")

In [297]:
stat, pval

(50.252984742109945, 1.0)

In our case, we have:

$$ 50.252984742109945 > -1.646$$

our statistics t is bigger than the critical value. Therefore, we **accept the null hypothesis** . Great!!

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [298]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [299]:
# Your code here: (Get the department which has most employees)
most_employees_department = salaries['JobTitle'].value_counts().idxmax()
print(f"The department with the most employees is: {most_employees_department}")



The department with the most employees is: Transit Operator


In [300]:
sample2 = salaries[salaries['JobTitle']== 'Transit Operator']['Hourly_Wage'].to_numpy()
sample2

array([9.712338e+01, 9.507354e+01, 9.364871e+01, ..., 1.254950e-01,
       9.471500e-02, 7.675000e-03])

In [301]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle"
# Computing manually our statistic.
t = (np.mean(sample2) - 35)/(np.std(sample2, ddof=1)/np.sqrt(len(sample2)))
t


19.064990324906383

In [302]:
# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here:
critical_value = scipy.stats.t.ppf(1-0.05, df=len(sample2)-1)
print("The critical value is {:.3f}".format(critical_value))

The critical value is 1.645


In our case, we have:

$$19.06499032490655 >  1.645$$

Therefore, we **reject the null hypothesis** !!

In [303]:
# Method 2: Use the p-value method.
# Your code here:
scipy.stats.t.cdf(t,df=len(sample2)-1)

1.0

In our case:

$$1 > 1 - \alpha = 0.95$$

Therefore, we **reject the null hypothesis**. Great!!


In [304]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
stat, pval = ttest_1samp(sample2, popmean = 35, alternative = "greater")
stat, pval

(19.064990324906383, 5.012771694204206e-76)

In this case the p_value is smaller than $\alpha=0.05$. Therefore we **reject the null hypothesis**

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [305]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage  
0  San Francisco     PT   255.366340  
1  San Francisco     FT   239.826105  
2  San Francisco     PT   218.112180  
3  San Francisco     FT   209.009610  
4  San Francisco     FT   208.717550

In [306]:
sample3 = salaries['Hourly_Wage'].to_numpy()

In [307]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:
confidence_level = 0.95
degrees_freedom = len(sample3) - 1  # or, for large values of sample size, just use that
sample_mean = np.mean(sample3)
# note that we use the standard error of the sample
# is an estimate of the standard error of the population (which is used in the theoretical formula)
sample_standard_error = scipy.stats.sem(sample3) # sem = standard error of the mean = std(mean)/sqrt(samplesize)

confidence_interval = scipy.stats.t.interval(confidence_level,
                                             degrees_freedom,
                                             sample_mean,
                                             sample_standard_error)

In [308]:
print("sample mean is {:.3f}".format(sample_mean))
print("confidence interval is ({:.3f},{:.3f})".format(*confidence_interval))


sample mean is 50.131
confidence interval is (49.798,50.463)


Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [309]:
# Your code here:
sample4 = salaries[salaries['JobTitle'].str.contains('Police')]['Hourly_Wage'].to_numpy()
sample4

array([2.09009610e+02, 1.50874905e+02, 1.44557475e+02, ...,
       1.06195000e-01, 6.73900000e-02, 3.51850000e-02])

In [310]:
# Method 1: Get the critical values which correspond to a 95% confidence.
# Your code here:
confidence_level = 0.95
degrees_freedom = len(sample4) - 1  # or, for large values of sample size, just use that
sample_mean = np.mean(sample4)
# note that we use the standard error of the sample
# is an estimate of the standard error of the population (which is used in the theoretical formula)
sample_standard_error = scipy.stats.sem(sample4) # sem = standard error of the mean = std(mean)/sqrt(samplesize)

confidence_interval = scipy.stats.t.interval(confidence_level,
                                             degrees_freedom,
                                             sample_mean,
                                             sample_standard_error)

In [311]:
print("sample mean is:", round(sample_mean, 2))
print("confidence interval is: ", confidence_interval)

sample mean is: 74.17
confidence interval is:  (72.95787917364453, 75.39116295155273)


# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [312]:
# Your code here: (Create the function)
def GroupOperation(x): 
    x = x.lower()
    if "police" in x:
        return "Policemen"
    elif "fire" in x:
        return "Firemen"
    elif "attorney" in x:
        return "Lawyer"
    elif "med" in x:
        return "Medical"
    else:
        return "Other"


In [313]:
# Your code here: (Apply the function) 
salaries['employee_group'] = salaries['JobTitle'].apply(GroupOperation) 

In [314]:
salaries.head()

Id          EmployeeName                  JobTitle    BasePay  \
0  110532           David Shinn            Deputy Chief 3  129150.01   
1  110533            Amy P Hart         Asst Med Examiner  318835.49   
2  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
3  110535        Gregory P Suhr           Chief of Police  307450.04   
4  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

   OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0         0.00  342802.63  38780.04  471952.64         510732.68  2014   
1     10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2         0.00   82313.70  96570.66  339653.70         436224.36  2014   
3         0.00   19266.72  91302.46  326716.76         418019.22  2014   
4         0.00   24165.44  91201.66  326233.44         417435.10  2014   

          Agency Status  Hourly_Wage employee_group  
0  San Francisco     PT   255.366340          Other  
1  San Francisco     FT   239.826105        Medical  
2  San Francisco     PT   218.112180          Other  
3  San Francisco     FT   209.009610      Policemen  
4  San Francisco     FT   208.717550        Firemen

Determine how many "PT" and "FT" employess have all the employees groups.

In [315]:
# Your code here: (Store the output dataframe into a new variable)

PT_FT = salaries.groupby(['employee_group', 'Status']).agg({'Status':'count'})
PT_FT

Status
employee_group Status        
Firemen        FT        1333
               PT         178
Lawyer         FT         317
               PT         102
Medical        FT         333
               PT         196
Other          FT       18821
               PT       14938
Policemen      FT        1530
               PT         371

In [316]:
group = salaries[['employee_group', 'Status']].pivot_table(index=['employee_group', 'Status'], aggfunc = {'employee_group':'count'} )
group 

employee_group
employee_group Status                
Firemen        FT                1333
               PT                 178
Lawyer         FT                 317
               PT                 102
Medical        FT                 333
               PT                 196
Other          FT               18821
               PT               14938
Policemen      FT                1530
               PT                 371

In [317]:
grouped_df = salaries.groupby(['employee_group', 'Status']).size().unstack(fill_value=0)
grouped_df

Status             FT     PT
employee_group              
Firemen          1333    178
Lawyer            317    102
Medical           333    196
Other           18821  14938
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [318]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:
# Total number of observations
N = len(salaries)

# Probabilities for each employee group
p_x = salaries['employee_group'].value_counts() / N

# Probabilities for each status
p_y = salaries['Status'].value_counts() / N

# Compute expected frequencies
expected_frequencies = pd.DataFrame(index=p_x.index, columns=p_y.index)
for x in p_x.index:
    for y in p_y.index:
        expected_frequencies.loc[x, y] = p_x[x] * p_y[y] * N

display(expected_frequencies)


FT           PT
Other      19779.46709  13979.53291
Policemen  1113.799785   787.200215
Firemen     885.297988   625.702012
Medical     309.942181   219.057819
Lawyer      245.492956   173.507044

In [323]:
N = len(salaries)

In [352]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"
firemen = (salaries['employee_group'] == 'Firemen').sum()
ft = (salaries['Status'] == 'FT').sum()

firemen_ft = (firemen / N) * (ft / N) * N
firemen_ft

885.2979878800596

885.2979878800596

In [361]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"
firemen = (salaries['employee_group'] == 'Firemen').sum()
pt = (salaries['Status'] == 'PT').sum()

firemen_pt = (firemen / N) * (pt / N) * N
firemen_pt

625.7020121199403

In [362]:
pt

15785

625.7020121199403

In [378]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"
lawyers = (salaries['employee_group'] == 'Lawyer').sum()
ft = (salaries['Status'] == 'FT').sum()

lawyers_ft = (lawyers / N) * (ft / N) * N
lawyers_ft

245.49295626852748

245.49295626852748

In [377]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"
lawyers = (salaries['employee_group'] == 'Lawyer').sum()
pt = (salaries['Status'] == 'PT').sum()
# pt = ((salaries['employee_group'] == 'Lawyer') & (salaries['Status'] == 'PT')).sum()


lawyers_pt = (lawyers / N) * (pt / N) * N
lawyers_pt

173.5070437314725

173.5070437314725

In [381]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"
medical = (salaries['employee_group'] == 'Medical').sum()
ft = (salaries['Status'] == 'FT').sum()

medical_ft = (medical / N) * (ft / N) * N
medical_ft

309.942181064561

2294.9783047823917

In [380]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"
medical = (salaries['employee_group'] == 'Medical').sum()
pt = (salaries['Status'] == 'PT').sum()

medical_pt = (medical / N) * (pt / N) * N
medical_pt

219.05781893543903

1622.021695217608

In [368]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"
other = (salaries['employee_group'] == 'Other').sum()
ft = (salaries['Status'] == 'FT').sum()

other_ft = (other / N) * (ft / N) * N
other_ft

19779.46708990267

17794.430966184842

In [383]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"
other = (salaries['employee_group'] == 'Other').sum()
pt = (salaries['Status'] == 'PT').sum()

other_pt = (other / N) * (pt / N) * N
other_pt

13979.532910097327

12576.569033815158

In [375]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"
policemen = (salaries['employee_group'] == 'Policemen').sum()
ft = (salaries['Status'] == 'FT').sum()

policemen_ft = (policemen / N) * (ft / N) * N
policemen_ft

1113.7997848841785

1113.7997848841785

In [374]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"
policemen = (salaries['employee_group'] == 'Policemen').sum()
pt = (salaries['Status'] == 'PT').sum()

policemen_pt = (policemen / N) * (pt / N) * N
policemen_pt

787.2002151158215

787.2002151158215

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [386]:
# Your code here:
ft_frequencies = [policemen_ft, lawyers_ft, medical_ft, other_ft, firemen_ft]
pt_frequencies = [policemen_pt, lawyers_pt, medical_pt, other_pt, firemen_pt]

In [387]:
frequencies_dict = {'FT': ft_frequencies, 'PT': pt_frequencies}

In [389]:
frequencies_df = pd.DataFrame(frequencies_dict)#index=salaries['employee_group'].unique())
frequencies_df

FT            PT
0   1113.799785    787.200215
1    245.492956    173.507044
2    309.942181    219.057819
3  13979.532910  13979.532910
4    885.297988    625.702012

FT            PT
0    885.297988    625.702012
1    245.492956    173.507044
2   2294.978305   1622.021695
3  17794.430966  12576.569034
4   1113.799785    787.200215

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [ ]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.


(2676.642333711905, 0.0, 4, array([[  885.29798788,   625.70201212],
        [  245.49295627,   173.50704373],
        [ 2294.97830478,  1622.02169522],
        [17794.43096618, 12576.56903382],
        [ 1113.79978488,   787.20021512]]))

Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

In [392]:
scipy.stats.chi2_contingency(grouped_df) 

Chi2ContingencyResult(statistic=1088.919545816174, pvalue=1.9093596261039512e-234, dof=4, expected_freq=array([[  885.29798788,   625.70201212],
       [  245.49295627,   173.50704373],
       [  309.94218106,   219.05781894],
       [19779.4670899 , 13979.5329101 ],
       [ 1113.79978488,   787.20021512]]))

based on the Chi-Square test result, there is a statistically significant relationship between the two categorical variables that I tested.